Incorporate dynamic dispatch into Messages as Interfaces article (#6538)
Particular/docs.particular.net: nservicebus/messaging/messages-as-interfaces.md

---
title: Interfaces as messages

> summary: Define messages as interfaces in NServiceBus to support multiple inheritance scenarios

Reviewer: ```suggestion
summary: Define interfaces as messages in NServiceBus to support multiple inheritance scenarios
```

---
title: Messages as Interfaces
summary: Define messages as interfaces in NServiceBus to support dynamic dispatch and polymorphic routing scenarios
component: core
reviewed: 2024-03-28
redirects:
- nservicebus/messages-as-interfaces
---

> NServiceBus supports using interfaces as messages to allow dynamic dispatch and polymorphic message routing scenarios. This makes systems more flexible and maintainable by encouraging messages and handlers to be decoupled from each other. It's useful for evolving systems to allow adding new features without affecting existing functionality. It's best described using an example.

Reviewer: ```suggestion
Using interfaces as messages enables multiple inheritance, where one message can inherit from multiple other messages. This in turn enables dynamic dispatch and polymorphic message routing scenarios. When used effectively to decouple messages and handlers from each other, systems are more flexible and maintainable. It's especially useful for evolving systems to allow adding new features without affecting existing functionality. It's best described using an example.
```

## Using NServiceBus for dynamic dispatch

> Consider the following handler stubs for a `UserCreated` message which inherits from the `IEvent` stub interface:

Reviewer: ```suggestion
For a common scenario involving the creation of a user as a standard system event, consider the following handler stubs for a `UserCreated` message which inherits from the `IEvent` stub interface:
```

(The text doesn't match the code right now, right? Or am I missing something?)

> ```csharp

Author: The code should be moved to snippets but I've left it as inline until this is reviewed

public class SaveUser : IHandleMessages<UserCreated>
{
    public async Task Handle(UserCreated message, IMessageHandlerContext) {
        // Do stuff when the user is created
    }
}

public class Audit : IHandleMessages<IMessage>
{
    public async Task Handle(IMessage message, IMessageHandlerContext context) {
        // Audit the message
    }
}
```

When a message of type `UserCreated` is published, both handlers will be invoked independently which in itself is useful. However, this idea becomes more powerful as the system evolves and more functionality is added.

> Assume there is a separate process that needs to happen if a user is created as part of a campaign. For example, say the subscriber must register statistics about the campaign. In NServiceBus, this can be done by defining a new event and handler:

Reviewer: ```suggestion
Meanwhile, imagine there is a separate process that needs to happen if a user is created as part of a campaign. For example, say the subscriber must register statistics about the campaign. In NServiceBus, this can be done by defining a new event and handler:
```

Does "campaign" mean something in .Net or something? I find it gets used a lot, but it means nothing to me.

Reviewer 2: Not that I'm aware of, to me it means an ad campaign, so it gets used a lot when you talk about getting messages from users, because it is a real life example on how someone will use the system.

Reviewer: Ok, so "ad campaign" does mean something to me.

```suggestion
Assume there is a separate process that needs to happen if a user is created as part of an advertising campaign. For example, say the subscriber must register statistics about the campaign. In NServiceBus, this can be done by defining a new event and handler:
```

```csharp
public class UserCreatedFromCampaign : UserCreated
{
    public Guid CampaignId { get; set;}
}

public class RecordStatistics : IHandleMessages<UserCreatedFromCampaign>
{
    public async Task Handle(UserCreatedFromCampaign, IMessageHandlerContext context) {
        // Record statistics
    }
}
```

> Here, when a `UserCreatedFromCampaign` message is published, the `RecordStatistics` handler is invoked as expected. But since `UserCreatedFromCampaign` inherits from `UserCreated`, the `SaveUser` handler is also invoked, along with the `Audit` handler (since `UserCreated` implements the `IMessage` stub interface). That is, new functionality has been added without having to change the existing handlers.

Reviewer 3: Suggestion: Add in a sentence that ties it back to the definition of "dynamic dispatch"

> Here, when a `UserCreatedFromCampaign` message is published, the `RecordStatistics` handler is invoked as expected. But since `UserCreatedFromCampaign` inherits from `UserCreated`, the `SaveUser` handler is also invoked, along with the `Audit` handler (since `UserCreated` implements the `IMessage` stub interface). That is, new functionality has been added without having to change the existing handlers.

Reviewer: ```suggestion
Here, when a `UserCreatedFromCampaign` message is published, the `RecordStatistics` handler is invoked as expected. But since `UserCreatedFromCampaign` inherits from `UserCreated`, the `SaveUser` handler is also invoked, along with the `Audit` handler (since `UserCreated` implements the `IMessage` stub interface). 

The existing handlers remain untouched, and the entirely new functionality is added with just a few lines of new code. This is the beauty of dynamic dispatch.
```

In this example, the new event, `UserCreatedFromCampaign`, has a clear relationship with `UserCreated`. However, it's possible that other events could be raised in relation to campaigns, not just users being created. This is where polymorphic routing can help.

## Multiple inheritance to support polymorphic routing

In the updated scenario, there are two concerns: a user is created and a campaign event occurred. The handlers for both events look as follows:

```csharp
public class SaveUser : IHandleMessages<UserCreated>
{
    public async Task Handle(UserCreated message, IMessageHandlerContext) {
        // Do stuff when the user is created
    }
}

public class RecordCampaignActivity : IHandleMessages<CampaignActivityOccurred>
{
    public async Task Handle(CampaignActivityOccurred message, IMessageHandlerContext) {

>         // Do stuff when the user is created

Reviewer 2: **Suggestion:** I think we should convey the different intent for each handler 
```suggestion
        // Do stuff when an event related to a campaign happened 
```

    }
}
```

Now assume the system should perform some action specifically when a user is created from a campaign. For example, perhaps an email is sent to the user thanking them for engaging. This can be accomplished with an event that inherits from both `UserCreated` and `CampaignActivityOccurred` as follows:

```csharp
public interface UserCreated : IEvent
{
    Guid UserId { get; set;}
}

public interface CampaignActivityOccurred : IEvent
{
    Guid CampaignId { get; set;}
}

public interface UserCreatedFromCampaign : UserCreated, CampaignActivityOccurred
{
}
```

When a user is created from a campaign, the publisher can publish the latter event. It will get handled by the `SaveUser` and `RecordCampaignActivity` handlers defined earlier, and perhaps another handler that explicitly handles `UserCreatedFromCampaign` events. Furthermore, other campaign events can be created that inherit from `CampaignActivityOccurred` without interfering with any functionality around creating users as part of a campaign.

> With this approach, events can support multiple inheritance and the system can create messages on the fly that implement these interfaces as described below.

Reviewer 3: Suggestion: Add in a sentence that ties it back to the definition of "polymorphic routing"

> With this approach, events can support multiple inheritance and the system can create messages on the fly that implement these interfaces as described below.

Reviewer 2: **Question:** You mean the interfaces defined before or after this paragraph?
```suggestion
With this approach, events can support multiple inheritance and the system can create messages on the fly that implement these interfaces as described above.
```

> With this approach, events can support multiple inheritance and the system can create messages on the fly that implement these interfaces as described above.

Reviewer: Something about this sentence doesn't feel right, but I'm not sure if I have anything better. I'm wondering if it's needed?

```suggestion
```
